### PR TITLE
Memoize some #hash calls

### PR DESCRIPTION
**Pull request by ngan**
## What was the end-user or developer problem that led to this PR?

`require 'bundler/setup` is slow for large Gemfiles (eg 500+ gems)

## What is your fix for the problem, implemented in this PR?

Memoize some `#hash` implementations since they don't change and we do it a lot for large gem files.

**Before**
```
➜  time ruby -e 'require "bundler/setup"' 
1.49s user 0.11s system 97% cpu 1.645 total
➜  time ruby -e 'require "bundler/setup"'
1.50s user 0.12s system 96% cpu 1.668 total
➜ time ruby -e 'require "bundler/setup"'
1.50s user 0.12s system 96% cpu 1.678 total
```

**After**
```
➜  time ruby -e 'require "bundler/setup"'
0.95s user 0.12s system 94% cpu 1.136 total
➜  time ruby -e 'require "bundler/setup"'
0.95s user 0.12s system 94% cpu 1.125 total
➜ time ruby -e 'require "bundler/setup"'
0.96s user 0.12s system 95% cpu 1.128 total
```

About 30% less time.

## Make sure the following tasks are checked

- [ ] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [ ] Write code to solve the problem
- [ ] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/lib/bundler/lazy_specification.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def eql?(other)
     end
 
     def hash
-      identifier.hash
+      @__hash ||= identifier.hash
```
Comment (ngan) [inline bundler/lib/bundler/lazy_specification.rb:38]: This is the money maker right here.

**Comment by ngan** [comment]
@deivid-rodriguez here's a small change for a pretty decent performance improvement. I tried memoizing more things in rubygems but decided not to since I'm unsure how useful it'd be.

**Review by deivid-rodriguez — COMMENTED** [review]
This looks very good, thanks for this. A 30% improvement for a one line change feels pretty great.

Do any of the other changes have any effect at all on your application? I'm asking because it might be best to just change "the money maker" here. Other changes involve vendored code from Molinillo and code monkey patched from RubyGems, so we probably should not change those without changing the upstream too. Also, I'm not sure changing `sources` has any effect given that the source list is usually short and sources are not compared often.

**Comment by deivid-rodriguez** [comment]
I tried this patch on `rails/rails` Gemfile and it leads to a pretty nice 8% `bundler/setup` time improvement! Also verified that only the change in `LazySpecification#hash` contributes to that improvement.

File: bundler/lib/bundler/lazy_specification.rb (hunk 1 above)
Comment (ngan) [reply bundler/lib/bundler/lazy_specification.rb:38]: Was following this pattern: https://github.com/rubygems/rubygems/blob/0aa81b10c8f6c20d03747400d4386514416d7244/bundler/lib/bundler/lazy_specification.rb#L116

**Comment by ngan** [comment]
Yeah, I got a little carried away. Happy to just memorize the lazy spec only. It’s late for me so I won’t be able to get to it until tomorrow. Feel free to modify this pr and land! It’s a very simple change. Otherwise I can get to it tomorrow.

**Comment by ngan** [comment]
Closing this out in favor of https://github.com/rubygems/rubygems/pull/5533

**Comment by deivid-rodriguez** [comment]
So, we can confirm that #5533 does beat the 30% improvement of this PR in your application, and that this patch no longer has an effect in performance after #5533?

**Comment by deivid-rodriguez** [comment]
Thank you so much for your investigation and patch, by the way!! ❤️ I obviously credited you on #5533 since I wouldn't have found the culprit without this hint!

**Comment by ngan** [comment]
No problem. @technicalpickles and I will find some more time and more hints for you 😄

There’s actually a problem with the latest released version though so we haven’t been able to upgrade. We’re unable to do a bundle install from scratch. I’ll take a closer look today to see if it’s something on our end or it’s because of the last speed up changes.

**Comment by ngan** [comment]
> Ouch, yeah, let me know what you find! :)

@deivid-rodriguez I started an issue: https://github.com/rubygems/rubygems/issues/5538

**Pull request closed without merging**